### PR TITLE
Compile time evaluation of PubSub Config

### PR DESCRIPTION
**Pull request by DomBlack**
This commit adds the ability for Encore's parser
to evaulate the configuration struct for both
Topics and Subscriptions as compile time constants.

It means the compiler can then attach all the configuration
option data into the metadata for the application, which will
then allow the backend to provision the infrastructure as required.

### CODE REVIEW

File: parser/est/est.go [hunk 1]
```diff
@@ -121,6 +121,12 @@ type PubSubSubscriber struct {
 	FuncFile *File        // The file the subscriber function is declared in
 	DeclFile *File        // The file that the subscriber is defined in
 	IdentAST *ast.Ident   // The AST node representing the value this topic is bound against
+
+	AckDeadline      int64 // time in nanoseconds
```
Comment (eandre) [inline parser/est/est.go:125]: Why are the `int64` as opposed to `time.Duration`?

File: proto/encore/parser/meta/v1/meta.proto [hunk 2]
```diff
@@ -207,8 +207,17 @@ message PubSubTopic {
...
+    string service_name      = 2; // The service that the subscriber is in
+    int64  ack_deadline      = 3; // How long has a consumer got to process and ack a message in nanoseconds
+    int64  message_retention = 4; // How long is an undelivered message kept in nanoseconds
+    RetryPolicy retry_policy = 5; // The retry policy for the subscription
+  }
+
+  message RetryPolicy {
+    int64 min_backoff = 1; // min backoff in nanoseconds
+    int64 max_backoff = 2; // max backoff in nanoseconds
+    int64 max_retries = 3; // max number of retires
```
Comment (eandre) [inline proto/encore/parser/meta/v1/meta.proto:220]:
```suggestion
    int64 max_retries = 3; // max number of retries
```

File: parser/literals.go [hunk 3]
```diff
@@ -75,12 +94,63 @@ elemLoop:
...
+		}
+
+	case *ast.BinaryExpr:
+		x := p.parseConstantValue(file, value.X)
+		y := p.parseConstantValue(file, value.Y)
+		return constant.BinaryOp(x, value.Op, y)
+
+	case *ast.UnaryExpr:
+		x := p.parseConstantValue(file, value.X)
+		fmt.Println("Uniary op", value.Op, x)
```
[the remark this answers is not in the document]
Comment (DomBlack) [reply parser/literals.go:138]: whoops 🙈

File: parser/literals.go [hunk 4]
```diff
@@ -75,22 +97,300 @@ elemLoop:
...
+			return constant.BinaryOp(lhs, value.Op, rhs)
+		case token.QUO:
+			// constant.BinaryOp panics when dividing by zero
+			if floatValue, _ := constant.Float64Val(constant.ToFloat(rhs)); floatValue <= 0.000000001 && floatValue >= -0.000000001 {
+				fmt.Printf("this is a float! %v\n", floatValue)
+				p.errf(value.Pos(), "cannot divide by zero")
+				return constant.MakeUnknown()
+			}
+
+			fmt.Println("Divide")
```
Comment (eandre) [inline parser/literals.go:160]: Remove

File: parser/literals.go [hunk 5]
```diff
@@ -75,22 +97,300 @@ elemLoop:
...
+			return constant.MakeUnknown()
+		}
+
+		switch value.Op {
+		case token.MUL, token.ADD, token.SUB, token.REM, token.AND, token.OR, token.XOR, token.AND_NOT:
+			return constant.BinaryOp(lhs, value.Op, rhs)
+		case token.QUO:
+			// constant.BinaryOp panics when dividing by zero
+			if floatValue, _ := constant.Float64Val(constant.ToFloat(rhs)); floatValue <= 0.000000001 && floatValue >= -0.000000001 {
+				fmt.Printf("this is a float! %v\n", floatValue)
```
Comment (eandre) [inline parser/literals.go:155]: Remove

**Pull request merged**